src/Mercury.ts

```diff
@@ -278,8 +169,41 @@ export class Mercury {
 		return this.emitOnFinishedCallback(eventId)
 	}
 
+	protected setAdapter(options: {
+		adapter: MercuryAdapterKind
+		connectionOptions: Record<string, any>
+	}): boolean {
+		const { adapter, connectionOptions } = options
+		log.debug('setAdapter', { options })
+
+		if (this.adapter) {
+			this.adapter.disconnect()
+			this.adapter = undefined
+		}
+
+		// TODO: Globby the adapters directory and set the correct one when we have multiple
+		let isAdapterSet = false
+		switch (adapter) {
+			case MercuryAdapterKind.SocketIO:
+				this.adapter = new MercuryAdapterSocketIO()
```

Comment:
A more extensible way to implement the adapter pattern would be to let me pass an instantiated adapter (which has a consistent `init` or `attach` method to do it's work) to this class (vs an enum and switch statement)

The way it is today, Mercury is technically; "not extensible", since you'd have to change core code every time you wanted to add an adapter.

Reply:
you would make the incoming type an interface that the abstract adapter extends (or you could type it as the abstract adapter, but I've seen it complain if you return `any` in the parent class and something more concrete, like a bool or string, from a subclass. In those cases typescript will say, "signatures are not compatible"

I think in this case, typing the incoming as the abstract adapter (assuming `init` is the reusable piece here).

Reply:
Agreed. The initial thinking with the adapters is that it would all be internal to Mercury and it would actually be up to the API after the initial connection call to tell the client what adapter to use. However, even if we do that we should still give the ability to pass in a custom adapter